internal/luks2/metadata.go

```diff
@@ -0,0 +1,874 @@
+// or file containing a LUKS2 volume with an integral header, or it can be a detached header file.
+// Data is interpreted in accordance with the LUKS2 On-Disk Format specification
+// (https://gitlab.com/cryptsetup/LUKS2-docs/blob/master/luks2_doc_wip.pdf).
+//
+// This function will verify the checksum of both the primary and secondary headers if found, and will
+// return the decoded form of one of the headers according to the following rules:
+//  - If both headers have valid checksums and the same sequence ID, return the primary header.
+//  - If both headers have valid checksums but different sequence IDs, return the newest header.
+//  - If only one header has a valid checksum, return that header.
+//
+// libcryptsetup performs some additional validation of the JSON metadata from both the primary
+// and secondary headers, and rejects a header if the JSON metadata isn't correctly formed. We don't
+// duplicate that logic here - we assume that anything that modifies the LUKS2 headers (which should
+// only be libcryptsetup) will write well-formed JSON metadata. Corruption of the JSON metadata outside
+// of modifications by libcryptsetup will be detected by the checksum verification.
+//
+// Note that this function does not attempt recovery of either header in the event that one of the
+// headers is not valid - we leave this to libcryptsetup, which happens automatically on any
+// cryptsetup or systemd-cryptsetup invocation. This package does not directly perform any
+// modifications to the header.
+//
+// This function requires an advisory shared lock on the LUKS container associated with the
+// specified path. If the mode parameter is LockModeBlocking, this function will block until the
+// lock can be obtained. If the mode parameter is LockModeNonBlocking, a wrapped syscall.Errno
+// error with the value of syscall.EWOULDBLOCK will be returned if the lock can not be obtained.
+func DecodeHdr(path string, lockMode LockMode) (*HdrInfo, error) {
+	releaseLock, err := acquireSharedLock(path, lockMode)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot acquire shared lock: %w", err)
+	}
+	defer releaseLock()
+
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	// Try to decode and check the primary header
+	primaryHdr, primaryJSON, primaryErr := decodeAndCheckHeader(f, 0, true)
+
+	var secondaryHdr *binaryHdr
+	var secondaryJSON []byte
+	if primaryErr != nil {
+		// No valid primary header. Try to decode and check a secondary header from one of the
+		// well known offsets.
+		for _, off := range []int64{0x4000, 0x8000, 0x10000, 0x20000, 0x40000, 0x80000, 0x100000, 0x200000, 0x400000} {
+			secondaryHdr, secondaryJSON, err = decodeAndCheckHeader(f, off, false)
+			if err == nil {
+				break
+			}
+		}
+	} else {
+		// Try to decode and check the secondary header immediately after the primary header.
+		secondaryHdr, secondaryJSON, _ = decodeAndCheckHeader(f, int64(primaryHdr.HdrSize), false)
```

Comment:
is  int64(primaryHdr.HdrSize)  and 0x4000 the same?

Reply:
I'm not sure I understand the question here.

Reply:
I reformulated the question again

Reply:
primaryHdr.HdrSize will be set to whatever the size of the header is set to, which can be any one of a finite set of values defined in the LUKS2 spec. The secondary header starts immediately after the end of the primary header. These offsets are the ones tried when the primary header is invalid, starting at 0x4000 (the minimum header size).